### PR TITLE
feat(storage): replace Tree generic with Arc<dyn TreeViewer>

### LINKED ISSUE #7615: Replace Tree generic with Arc<dyn TreeViewer>
### Describe the feature

This generic doesn't have any performance benefits because it's just just for simple get calls, which will also be refactored shortly.

It would make the setup a lot more convenient if we'd use `Arc< dyn TreeViewer>` here instead. Treeviewer trait is object safe so this should be very simple.

https://github.com/paradigmxyz/reth/blob/041e29347b2a01b9e971ebea988561f4dae13f30/crates/storage/provider/src/providers/mod.rs#L74-L75

This would let us cut down on a bunch of bloated types in the node-builder

https://github.com/paradigmxyz/reth/blob/041e29347b2a01b9e971ebea988561f4dae13f30/crates/node-builder/src/builder.rs#L63-L64

and allows us to be less constraint during the node-builder setup

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by KyrylR**
Fixes #7615 

This PR introduces a new helper trait `TreeViewer`, that combines all traits that are needed for the `tree` field inside the `BlockchainProvider` structure.

### CODE REVIEW

File: crates/consensus/beacon/src/engine/test_utils.rs [hunk 1]
```diff
@@ -90,7 +84,7 @@ impl<DB> TestEnv<DB> {
         loop {
             let result = self.send_new_payload(payload.clone(), cancun_fields.clone()).await?;
             if !result.is_syncing() {
-                return Ok(result)
+                return Ok(result);
```
Comment (joshieDo) [inline crates/consensus/beacon/src/engine/test_utils.rs:87]: please run nightly fmt instead

File: crates/node-builder/src/builder.rs [hunk 2]
```diff
@@ -62,11 +62,9 @@ use tokio::sync::{mpsc::unbounded_channel, oneshot};
 
 /// The builtin provider type of the reth node.
 // Note: we need to hardcode this because custom components might depend on it in associated types.
-type RethFullProviderType<DB, Evm> =
-    BlockchainProvider<DB, ShareableBlockchainTree<DB, EvmProcessorFactory<Evm>>>;
+type RethFullProviderType<DB> = BlockchainProvider<DB>;
 
-type RethFullAdapter<DB, N> =
-    FullNodeTypesAdapter<N, DB, RethFullProviderType<DB, <N as NodeTypes>::Evm>>;
+type RethFullAdapter<DB, N> = FullNodeTypesAdapter<N, DB, RethFullProviderType<DB>>;
```
Comment (mattsse) [inline crates/node-builder/src/builder.rs:67]: nice!

File: crates/blockchain-tree/src/shareable.rs [hunk 3]
```diff
@@ -206,3 +206,10 @@ where
         self.tree.read().subscribe_canon_state()
     }
 }
+
+impl<DB, EF> TreeViewer for ShareableBlockchainTree<DB, EF>
+where
+    DB: Database + Clone,
+    EF: ExecutorFactory,
+{
+}
```
Comment (mattsse) [inline crates/blockchain-tree/src/shareable.rs:215]: we don't need this if we add a blanket impl instead

File: crates/storage/provider/src/traits/tree_viewer.rs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+/// Helper trait to combine all the traits we need for the BlockchainProvider
+///
+/// This is a temporary solution
+pub trait TreeViewer:
+    BlockchainTreeViewer
+    + BlockchainTreePendingStateProvider
+    + CanonStateSubscriptions
+    + BlockchainTreeEngine
+{
+}
```
Comment (mattsse) [inline crates/storage/provider/src/traits/tree_viewer.rs:14]: add blanket impl for T: where T satisfies all those traits

**Review by mattsse — CHANGES_REQUESTED** [review]
one more suggestion

**Review by mattsse — APPROVED** [review]
sweet!

**Review by joshieDo — APPROVED** [review]

**Pull request merged**